File: android/heading-core/src/main/kotlin/com/fengshuicompass/headingcore/config/SharedArtifacts.kt
package com.fengshuicompass.headingcore.config

import java.io.File

/**
 * Locates the repository-root shared artifacts (`config/`, `schemas/`, `testdata/`)
 * that the Android runtime, the iOS runtime and `analysis/` all read.
 *
 * SPEC.md §37.1 requires both platforms to consume the *same* files rather than copies,
 * because independently translated constants are a common source of false parity. In a
 * shipped app these files are bundled read-only; in the JVM test runtime they are read
 * from the checkout via the `fsc.repoRoot` system property set by the module build.
 */
public object SharedArtifacts {

    public const val REPO_ROOT_PROPERTY: String = "fsc.repoRoot"

    public val repoRoot: File
        get() {
            val configured = System.getProperty(REPO_ROOT_PROPERTY)
                ?: error(
                    "System property $REPO_ROOT_PROPERTY is not set. The Gradle test task " +
                        "sets it to the repository root; a runtime that cannot locate the " +
                        "shared artifacts must fail rather than fall back to a copy."
                )
            val root = File(configured)
            require(root.isDirectory) { "$REPO_ROOT_PROPERTY=$configured is not a directory" }
            return root
        }

    public val precisionProfileFile: File get() = resolve("config/precision-profile-v1.json")

    public val fengShuiRuleSetFile: File get() = resolve("config/feng-shui-rules-v1.json")

    public val gradeReachabilityClaimsFile: File
        get() = resolve("testdata/grade-reachability-claims-v1.json")

    public val exampleEngineOutputEventFile: File
        get() = resolve("testdata/telemetry-event-engine-output-v1.example.json")

    // ---------------------------------------------------------------------------------
    // The frozen `fixtures-v1` shared contract (SPEC.md §37.1).
    //
    // "give both agents the same files and tolerances, with neither editing shared fixtures
    // unilaterally; a contract change requires one reviewed change set, regenerated fixtures,
    // both test suites, and a new fixture version".
    // ---------------------------------------------------------------------------------

    public const val FIXTURE_VERSION: String = "fixtures-v1"

    public val circularMathFixture: File get() = resolve("testdata/angles/circular-math-v1.json")

    public val estimatorsFixture: File get() = resolve("testdata/angles/estimators-v1.json")

    public val circularAggregateFixture: File
        get() = resolve("testdata/angles/circular-aggregate-v1.json")

    public val attitudeGoldenFixture: File
        get() = resolve("testdata/quaternions/attitude-golden-v1.json")

    public val fengShuiClassificationFixture: File
        get() = resolve("testdata/fengshui/classification-v1.json")

    public val fengShuiReferenceTransformFixture: File
        get() = resolve("testdata/fengshui/reference-transform-v1.json")

    public val uncertaintyCompositionFixture: File
        get() = resolve("testdata/uncertainty/composition-v1.json")

    public val magneticClassificationFixture: File
        get() = resolve("testdata/magnetic/classification-v1.json")

    public val referenceResolutionFixture: File
        get() = resolve("testdata/reference/resolution-v1.json")

    public val stateReducerFixture: File get() = resolve("testdata/state/reducer-v1.json")

    public val certificationKeyFixture: File get() = resolve("testdata/certification/key-v1.json")

    public val telemetryCodecFixture: File get() = resolve("testdata/telemetry/codec-v1.json")

    public val fixturesManifest: File get() = resolve("testdata/fixtures-v1.manifest.json")

    public fun resolve(relativePath: String): File {
        val file = File(repoRoot, relativePath)
        require(file.isFile) { "required shared artifact is missing: $relativePath" }
        return file
    }
}

File: ios/HeadingCore/Sources/HeadingCore/SharedArtifacts.swift
import Foundation

/// Locates the repository-root shared artifacts (`config/`, `schemas/`, `testdata/`) that
/// the Android runtime, the iOS runtime and `analysis/` all read.
///
/// SPEC.md §37.1 requires both platforms to consume the *same* files rather than copies,
/// because independently translated constants are a common source of false parity. In a
/// shipped app these files are bundled read-only; in the test runtime they are read from
/// the checkout, located by walking up from this source file.
public enum SharedArtifacts {

    public enum ArtifactError: Error, CustomStringConvertible {
        case repositoryRootNotFound(searchedFrom: String)
        case missingArtifact(String)

        public var description: String {
            switch self {
            case .repositoryRootNotFound(let from):
                return "could not locate the repository root walking up from \(from); a runtime "
                    + "that cannot find the shared artifacts must fail rather than fall back to a copy"
            case .missingArtifact(let path):
                return "required shared artifact is missing: \(path)"
            }
        }
    }

    /// Environment override, used by CI where the checkout is not the source location.
    public static let repoRootEnvironmentVariable = "FSC_REPO_ROOT"

    public static func repositoryRoot(from file: StaticString = #filePath) throws -> URL {
        if let override = ProcessInfo.processInfo.environment[repoRootEnvironmentVariable] {
            return URL(fileURLWithPath: override, isDirectory: true)
        }
        var directory = URL(fileURLWithPath: "\(file)").deletingLastPathComponent()
        let marker = "config/precision-profile-v1.json"
        for _ in 0..<12 {
            if FileManager.default.fileExists(atPath: directory.appendingPathComponent(marker).path) {
                return directory
            }
            directory = directory.deletingLastPathComponent()
        }
        throw ArtifactError.repositoryRootNotFound(searchedFrom: "\(file)")
    }

    public static func url(_ relativePath: String, from file: StaticString = #filePath) throws -> URL {
        let candidate = try repositoryRoot(from: file).appendingPathComponent(relativePath)
        guard FileManager.default.fileExists(atPath: candidate.path) else {
            throw ArtifactError.missingArtifact(relativePath)
        }
        return candidate
    }

    public static func precisionProfileURL(from file: StaticString = #filePath) throws -> URL {
        try url("config/precision-profile-v1.json", from: file)
    }

    public static func fengShuiRuleSetURL(from file: StaticString = #filePath) throws -> URL {
        try url("config/feng-shui-rules-v1.json", from: file)
    }

    public static func gradeReachabilityClaimsURL(from file: StaticString = #filePath) throws -> URL {
        try url("testdata/grade-reachability-claims-v1.json", from: file)
    }

    public static func exampleEngineOutputEventURL(from file: StaticString = #filePath) throws -> URL {
        try url("testdata/telemetry-event-engine-output-v1.example.json", from: file)
    }
}
